tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -50,18 +54,70 @@ private SpanMessagePackFormatter()
+        public int Serialize(ref byte[] bytes, int offset, ArraySegment<Span> value, IFormatterResolver formatterResolver)
+        {
+            var traceContext = value.Array![value.Offset].Context.TraceContext;
+            var traceChunk = new TraceChunkModel(value, traceContext);
+            return Serialize(ref bytes, offset, traceChunk, formatterResolver);
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, TraceChunkModel value, IFormatterResolver formatterResolver)
+        {
+            int originalOffset = offset;
+            var spans = value.Spans;
+            var spanIds = new HashSet<ulong>(spans.Count);
+
+            // Using a for loop to avoid the boxing allocation on ArraySegment.GetEnumerator
+            for (var i = 0; i < spans.Count; i++)
+            {
+                // get all span ids for quick lookup below
+                var span = spans.Array![i + spans.Offset];
+                spanIds.Add(span.SpanId);
+            }
+
+            // start writing span[]
+            offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, spans.Count);
+
+            // Using a for loop to avoid the boxing allocation on ArraySegment.GetEnumerator
+            for (var i = 0; i < spans.Count; i++)
+            {
+                var span = spans.Array![i + spans.Offset];
+                var parentSpanId = span.Context.ParentId ?? 0;
+                bool isRoot = parentSpanId == 0 || span == span.Context.TraceContext.RootSpan;
+                bool isOrphan = parentSpanId > 0 && !spanIds.Contains(parentSpanId);
+                bool isFirst = i == 0;
+
+                var spanModel = new SpanModel(span, value, isRoot, isOrphan, isFirst);
+                offset += Serialize(ref bytes, offset, spanModel, formatterResolver);
+            }
+
+            return offset - originalOffset;
+        }
+
+        // IMessagePackFormatter<Span> is deprecated, use IMessagePackFormatter<SpanModel>
         public int Serialize(ref byte[] bytes, int offset, Span value, IFormatterResolver formatterResolver)
         {
+            // without more context, we can only determine if this span is a local root
+            bool isRoot = value.Context.ParentId is null or 0 || value == value.Context.TraceContext.RootSpan;
+            var spanModel = new SpanModel(value, traceChunk: null, isRoot, isChunkOrphan: false, isFirstSpanInChunk: false);
+            return Serialize(ref bytes, offset, spanModel, formatterResolver);
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, SpanModel value, IFormatterResolver formatterResolver)
+        {
+            var span = value.Span;
+
             // First, pack array length (or map length).
             // It should be the number of members of the object to be serialized.
             var len = 8;
 
-            if (value.Context.ParentId != null)
+            if (span.Context.ParentId > 0)
```

Comment:
Is this the same as: 
```suggestion
            if (!value.IsLocalRoot)
```

Is there any reason for using one over the other? I think `IsLocalRoot` is clearer?

EDIT: I guess this begs another question, is `IsLocalRoot` the right name given this implies it's a true root (not just a local root) if it has `parentId == 0`?

Reply:
It is not the same because a local root span can have a non-zero parent id if we received a trace id and span id from a remote service upstream.

This change is not related to the refactor. It just seemed like an easy way to save a few bytes in the rare case where we have parent id of `0` instead of `null` (the trace agent treats them equally).

Reply:
> It is not the same because a local root span can have a non-zero parent id if we received a trace id and span id from a remote service upstream.

Ok, but in the `Serialize` methods that create a `SpanModel` we're doing:

```csharp
bool isRoot = value.Context.ParentId is null or 0 || value == value.Context.TraceContext.RootSpan;
```

And setting `isRoot` as `IsLocalRoot` in the `SpanModel`. But afaict, we're _not_ calculating local root right? Seems like something needs updating - is that just a TODO? 🙂

Reply:
In the code you quoted above, I was trying to use a shortcut: if `parent id is null or 0`, then we know this span is a root span (in both the local and the distributed sense), so there's no need to compare it to `TraceContext.RootSpan`.

It should always have the same results as
```csharp
bool isRoot = value == value.Context.TraceContext.RootSpan;
```

Reply:
I keep re-reading and I'm still not sure where the confusion lies. Maybe it's this: at this point we're not trying to decide if this is a root span or not. We just want to know if this fields has a positive non-zero value so we can send the value to the agent. If the value is null or zero, we can skip it and save a few bytes.

(actually null used to throw an error in the agent, so we _had_ to skip it, I don't know if that's fixed)

Reply:
>  keep re-reading and I'm still not sure where the confusion lies.

I don't have any confusion at _this_ point any more 🙂 

My previous question was whether the boolean we're passing to `SpanModel` was _actually_ calculating the _root_ or the _local root_. I repeatedly misread this previously, so I see it's definitely calculating the local root (and you renamed it as such).

Thanks for entertaining my denseness! 😄